optuna/storages/redis.py

```diff
@@ -0,0 +1,586 @@
+        else:
+            return {}
+
+    def _set_study_param_distribution(self, study_id, param_distribution):
+        # type: (int, Dict) -> None
+
+        self._redis.set(
+            self._key_study_param_distribution(study_id), pickle.dumps(param_distribution)
+        )
+
+    def get_all_study_summaries(self):
+        # type: () -> List[structs.StudySummary]
+
+        study_summaries = []
+        study_ids = [pickle.loads(sid) for sid in self._redis.lrange("study_list", 0, -1)]
+        for study_id in study_ids:
+            study_summary = self._get_study_summary(study_id)
+            study_summaries.append(study_summary)
+
+        return study_summaries
+
+    def create_new_trial(self, study_id, template_trial=None):
+        # type: (int, Optional[structs.FrozenTrial]) -> int
+
+        self._check_study_id(study_id)
+
+        if template_trial is None:
+            trial = self._create_running_trial()
+        else:
+            trial = copy.deepcopy(template_trial)
+
+        if not self._redis.exists("trial_counter"):
+            self._redis.set("trial_counter", -1)
+
+        trial_id = self._redis.incr("trial_counter", 1)
+        trial_number = self._redis.incr("study_id:{:010d}:trial_number".format(study_id))
+        trial.number = trial_number
+        trial._trial_id = trial_id
+
+        with self._redis.pipeline() as pipe:
+            pipe.multi()
+            pipe.set(self._key_trial(trial_id), pickle.dumps(trial))
+            pipe.set("trial_id:{:010d}:study_id".format(trial_id), pickle.dumps(study_id))
+            pipe.rpush("study_id:{:010d}:trial_list".format(study_id), trial_id)
+            pipe.execute()
+
+            pipe.multi()
+            study_summary = self._get_study_summary(study_id)
+            study_summary.n_trials = len(self._get_study_trials(study_id))
+            min_datetime_start = min([t.datetime_start for t in self.get_all_trials(study_id)])
+            study_summary.datetime_start = min_datetime_start
+            pipe.set(self._key_study_summary(study_id), pickle.dumps(study_summary))
+            pipe.execute()
+
+        return trial_id
+
+    @staticmethod
+    def _create_running_trial():
+        # type: () -> structs.FrozenTrial
+
+        return structs.FrozenTrial(
+            trial_id=-1,  # dummy value.
+            number=-1,  # dummy value.
+            state=structs.TrialState.RUNNING,
+            params={},
+            distributions={},
+            user_attrs={},
+            system_attrs={},
+            value=None,
+            intermediate_values={},
+            datetime_start=datetime.now(),
+            datetime_complete=None,
+        )
+
+    def set_trial_state(self, trial_id, state):
+        # type: (int, structs.TrialState) -> bool
+
+        trial = self.get_trial(trial_id)
+        self.check_trial_is_updatable(trial_id, trial.state)
+
+        if state == structs.TrialState.RUNNING and trial.state != structs.TrialState.WAITING:
```

Comment:
I think that this condition could be violated if, for example, the following scenario occurred.

- Assumption:
  - We are executing a distributed optimization with two worker processes, and
  - there is a trial with ID `10` and state `TrialState.WAITING`.
- (1) "A" calls `set_trial_state(10, TrialState.RUNNING)`, and is suspended just before line 354.
- (2) "B" calls `set_trial_state(10, TrialState.RUNNING)`, and is also suspended just before line 354.
  - Note that at the this time, both "A" and "B" have the exactly same `trial` .
- (3) "A" is resumed. Since the condition (line 354) is evaluated to `False`, this process sets the trial state to `RUNNING`.
- (4) "A" continues the evaluation of the trial, and eventually sets the state to `COMPLETE`.
- (5) After that, "B" is resumed. Since this process has the same `trial` as "A" in the step `(3)`, the condition is evaluated to `False`.
   - Then "B" changes the trial state to `RUNNING` from `COMPLETE` (not `WAITING`).
   - => This is unexpected behavior.

Obviously this is an artificial scenario, but I think that this and other methods that update DB state have potential risks like this.

Reply:
To fix the above issue, it seems we need to rewrite the update methods to use `WATCH` redis command (i.e., optimistic locking mechanism) and retry if the watching keys were updated by other processes during a transaction. 
FYI, [The redis official doc about transaction](https://redis.io/topics/transactions) describes a typical flow to enable optimistic locking in Redis (the comments were added by me):
```
# EXAMPLE: An own implementation of INCR command (quoted from the redis official doc).
WATCH mykey     # (1) Watch target key
val = GET mykey # (2) Get the content of the key
val = val + 1   # (3) Update the content locally
MULTI           # (4) Start transaction
SET mykey $val  # (5) Enqueue the update request in the local command queue
EXEC            # (6) Execute the transaction atomically (if it failed, we need to retry)
```

Although we need to use the optimistic locking to fully support Optuna's distributed optimization, it could be a task requiring a bit large code modification. As the current implementation works fine when there is only single worker, I think that it's okay to merge this PR without distributed optimization support and address this issue in another PR.
What do you think of it?

Reply:
for 1) and 2) I do not understand why 2 processes will end up having the same trial ID = 10. 
line #308 of redis.py shows that the trial id is given by an atomic incr operation. This guarantees that calling create_new_trial will give you a different ID every time. (sequential integers). Perhaps there is a scenario that I do not currently see, where you reuse trials perhaps...

Also checking on rdb storage.py https://github.com/optuna/optuna/blob/master/optuna/storages/rdb/storage.py#L495-L497
has the same logic.

Perhaps is this is an issue we can merge this PR first. 
Then submit another PR tackling this concurrency scenario you describe having an integration test, but for all stores at once.

Reply:
> for 1) and 2) I do not understand why 2 processes will end up having the same trial ID = 10.

Sorry for the lack of explanation. The scenario could occur when a user uses [Study.enqueue_trial](https://optuna.readthedocs.io/en/latest/reference/study.html#optuna.study.Study.enqueue_trial) feature. To pop a queued trial (i.e., a trial with `WAITING` state) at the begging of a new trial round, multiple workers could call `set_trial_state` with the same trial ID simultaneously (see [here](https://github.com/optuna/optuna/blob/034cfdf1ff9cddcd9c0e5301f9a476181c5f2b41/optuna/study.py#L639) for the code).
Besides, though the detail has not been decided, we have a plan to expose `set_trial_state` to users (https://github.com/optuna/optuna/issues/685#issuecomment-554159519) via `Study` object. If it becomes a public API, users might call the method with the same ID from multiple processes.

> Perhaps is this is an issue we can merge this PR first.

Good. I'll take a final look this week and then approve this PR.

Reply:
Good to know. Thanks for the clarification.